### PR TITLE
Attachments: files can be attached for ticket and comments

### PR DESCRIPTION
**Pull request by nikolay-zakharov**

### CODE REVIEW

**Comment by xaviqv** [comment]
I've tested you pull request and it works great! For me it is a must for this project.

I have some suggestions of improvements for attached files:

**In ticket list**
- To have an icon indicating tickets with attached files. Also with the attached files count.
- An extra to the last point would be to be able to get the files directly form the ticket list. For this, I've got in mind two approaches: A kind of bootstrap based [dropdown menu](https://www.w3schools.com/bootstrap/bootstrap_dropdowns.asp) or a bootstrap [popover](https://www.w3schools.com/bootstrap/bootstrap_popover.asp) or using the [Datatables child rows](https://datatables.net/examples/api/row_details.html).

**In the ticket or comment form**
- After selecting many files, ability to discard them individually, like scripts [like jQuery file upload do](https://i.stack.imgur.com/qLbyQ.png).

**In the ticket card**
- Ability to discard already attached files on ticket or comment
- To have a list of all the ticket attached files near the ticket fields at the beggining.
- To view thumbnails of the attached files that are images. In this case, open the image directy instead of downloading it

**Related settings**
- Setting for allowed file extensions
- Setting for max number of attached files for each post (even new ticket or comment).
- Maybe another setting for the whole ticket attached files count? Although it doesn't make sense to me
- Setting for max file size

File: src/Controllers/TicketsController.php [hunk 1]
```diff
@@ -177,28 +183,54 @@ public function store(Request $request)
...
+        /** @var Attachment $attachment */
+        $attachment = Attachment::query()
+            ->where('id', $attachment_id)
+            ->whereHas('ticket', function ($ticketQuery) use ($user) {
+                // Ensure user has permissions to access the ticket
+
+                if ($user->isAdmin()) {
+                    // No restriction for admin
+                } elseif ($user->isAgent()) {
+                    $ticketQuery->agentUserTickets($user->id);
```
Comment (thekordy) [inline src/Controllers/TicketsController.php:223]: I think this should be related to the configured option 'agent_restrict', if agents are not restricted, let them download any file.

**Comment by thekordy** [comment]
@nikolay-zakharov I like it, well done :+1:  

I must say I agree with @xaviqv mentioned Related settings:
>>Setting for allowed file extensions
Setting for max number of attached files for each post (even new ticket or comment).
Maybe another setting for the whole ticket attached files count? Although it doesn't make sense to me
Setting for max file size

I am approving this PR, to make it available for you and others to work on these points.

**Review by thekordy — CHANGES_REQUESTED** [review]

File: src/Migrations/2017_03_21_144731_add_ticketit_attachments.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('ticketit_attachments', function (Blueprint $table) {
+            $table->increments('id');
+            $table->unsignedInteger('ticket_id');
+            $table->unsignedInteger('comment_id')->nullable();
```
Comment (thekordy) [inline src/Migrations/2017_03_21_144731_add_ticketit_attachments.php:18]: Fine by me, but I think if you convert this to a polymorphic relationship, will make your model reusable. just a thought.

**Pull request closed without merging**